src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/Documents/RunningDocumentTableEventSink.cs

```diff
@@ -2,51 +2,83 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Razor;
 using Microsoft.VisualStudio.Shell.Interop;
 
 namespace Microsoft.VisualStudio.Editor.Razor.Documents
 {
     internal class RunningDocumentTableEventSink : IVsRunningDocTableEvents3
     {
         private readonly VisualStudioEditorDocumentManager _documentManager;
+        private readonly ForegroundDispatcher _foregroundDispatcher;
 
-        public RunningDocumentTableEventSink(VisualStudioEditorDocumentManager documentManager)
+        public RunningDocumentTableEventSink(VisualStudioEditorDocumentManager documentManager, ForegroundDispatcher foregroundDispatcher)
         {
-            if (documentManager == null)
+            if (documentManager is null)
             {
                 throw new ArgumentNullException(nameof(documentManager));
             }
 
+            if (foregroundDispatcher is null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
             _documentManager = documentManager;
+            _foregroundDispatcher = foregroundDispatcher;
         }
 
         public int OnAfterAttributeChangeEx(uint docCookie, uint grfAttribs, IVsHierarchy pHierOld, uint itemidOld, string pszMkDocumentOld, IVsHierarchy pHierNew, uint itemidNew, string pszMkDocumentNew)
         {
-            // Document has been initialized.
-            if ((grfAttribs & (uint)__VSRDTATTRIB3.RDTA_DocumentInitialized) != 0)
-            {
-                _documentManager.DocumentOpened(docCookie);
-            }
+            _ = Task.Factory.StartNew(() => OnAfterAttributeChangeExAsync(
+                docCookie, grfAttribs, pszMkDocumentOld, pszMkDocumentNew),
+                CancellationToken.None,
+                TaskCreationOptions.None,
+                TaskScheduler.Default);
+            return VSConstants.S_OK;
+        }
 
-            if ((grfAttribs & (uint)__VSRDTATTRIB.RDTA_MkDocument) != 0)
+        private async Task OnAfterAttributeChangeExAsync(uint docCookie, uint grfAttribs, string pszMkDocumentOld, string pszMkDocumentNew)
+        {
+            await _foregroundDispatcher.RunOnForegroundAsync(async () =>
```

Comment:
Might be able to get away with having this just be on the UI thread. I'd be curious to know the reason why DocumentOpenedAynsc required the fake foreground thread